### ISSUE TITLE
Difference between `makeOpamRepo` and `makeOpamRepoRec`

### ISSUE THREAD

**Issue by sir4ur0n** [balsoft/opam-nix#97]
**Describe the bug**
The documentation states:
> Traverse a directory (**recursively in case of `makeOpamRepoRec`**), looking for opam files and collecting them into a repository in a format understood by opam.

However it seems that even `makeOpamRepo` traverses recursively.

My setup:
```
.
├── a.opam
├── b.opam
├── opam
│   ├── c.opam
│   ├── d.opam
```

So I expect `(makeOpamRepo ./.).passthru.sourceMap` to only include `a` and `b`, while I expect `(makeOpamRepoRec ./.).passthru.sourceMap` to include `a`, `b`, `c` and `d`.

Am I understanding it wrong?

Is there a way to only take the opam files at the given level, without exploring all sub-paths?
This is also problematic when we have a `_opam` directory (e.g. when not using `opam-nix`) as `makeOpamRepo` also fetches all opam files in that directory.

**Environment**
 - Version of the code: `master`

**Comment by balsoft** [comment]
Oh, yes, this is indeed a bug. The problem is this:

```nix
      contents = readDir dir;
      files = if recursive then
        readDirRecursive dir
      else
        (contents // optionalAttrs (contents.opam or null == "directory") {
          opam = readDir "${dir}/opam";
        });
```

So, it reads the `opam` subdirectory specifically. I'm not sure what the exact logic should be here, I suppose I'll go look at `opam`'s documentation :)

**Comment by sir4ur0n** [comment]
By looking at https://github.com/tweag/opam-nix/blob/main/src/opam.nix#L214 it seems that if there is a directory `opam/` present, it is always read, no matter if we call the recursive function or not.

I suggest at the very least documenting this exception.
IMHO it should also be removed.

Is there a way to call this function while ignoring the content of the `opam/` directory?
We use that directory in many repos to store the cross-compilation flavors of our opam repositories (e.g. `./a.opam` vs `./opam/a-android.opam`) so we would like to not have to change all our repo structures :pray:

**Comment by balsoft** [comment]
I'm not sure of the reason for this exception myself. I guess I was under the impression that upstream opam was reading that directory when doing `opam install`. I'll see if that's true.

**Comment by balsoft** [comment]
>  We use that directory in many repos to store the cross-compilation flavors of our opam repositories (e.g. ./a.opam vs ./opam/a-android.opam) so we would like to not have to change all our repo structures 🙏

Just out of interest, how do you then select the flavor you're building?

**Comment by balsoft** [comment]
Apparently, this behavior was important enough for a separate commit: ac7ba6e749c01379c748eda43ae715e1a8fa671f

**Comment by sir4ur0n** [comment]
> Just out of interest, how do you then select the flavor you're building?

Our cross-compilation CI replaces `<x>.opam` with `opam/<x>-<cross-compilation target>.opam`.

One of the reasons to keep cross-compilation opam files in a separate directory is that (non-opam-nix) developers often run `opam install --yes --deps-only --with-test ./*.opam` to setup a dev env, thus having the cross-compilation opam files at the root would blow up.

> Apparently, this behavior was important enough for a separate commit: https://github.com/tweag/opam-nix/commit/ac7ba6e749c01379c748eda43ae715e1a8fa671f

I don't mind in itself this default, but I would appreciate if we can bypass it somehow :sweat_smile: (and maybe document it to save time for future users :grin: )

FWIW I think we would be ok with passing the explicit list of opam files rather than using a function that discovers them

**Comment by balsoft** [comment]
Ok, so what I found is that `opam pin .` will also look into `opam/` subdirectory, just like `makeOpamRepo`. I think I'd better stick to that behavior for the main `makeOpamRepo` function; what I'll do is split it into two functions, one for discovering the package files and another for taking those package files and making a repo from them.

**Comment by sir4ur0n** [comment]
> what I'll do is split it into two functions, one for discovering the package files and another for taking those package files and making a repo from them.

Sounds great, that way, one could pass its own function (e.g. to hard-code the list of opam files)? If so, perfect for us :+1:

**Comment by balsoft** [comment]
> Sounds great, that way, one could pass its own function (e.g. to hard-code the list of opam files)? If so, perfect for us 👍

Yep, or just `findOpamFiles (readDir ./.)` to get the list of opam files in only one directory.

**Comment by balsoft** [comment]
Implemented in 94e511fb4fe36c248d188f74bcb24d5bb9411fe9 . Please close this issue once you get a chance to confirm that it works.

**Comment by sir4ur0n** [comment]
It seems to work fine for us with the latest `main`, thank you! :pray:
